### PR TITLE
fix world.is_resource_added() panicing

### LINKED ISSUE #2828: app.world().is_resource_added::<T>() panics instead of returning false
## Bevy version

0.5

## Operating system & version

Arch(tho not relevant)

## What you did

Calling `app.world().is_resource_added::<T>()` on a resource that wasn't added.

## What you expected to happen

function to return false

## What actually happened

It paniced

## Additional information

Crashed at this line

```
bevy_ecs/src/world/mod.rs:574:79
```

because of the `unwrap()` at the end.

### PR DESCRIPTION
**Pull request by RustyStriker**
# Objective

Fixes #2828

## Solution

replaced unwrap with match(and added a test to make sure everything works without crashing)

### CODE REVIEW

**Comment by MinerSebas** [comment]
Could you also remove the unwraps inside `is_resource_changed`.

**Comment by RustyStriker** [comment]
> Could you also remove the unwraps inside `is_resource_changed`.

if a resource doesn't exist, would you say it is never changed? or indeed should we crash the program because the user might try to access an unregistered resource?

File: crates/bevy_ecs/src/world/mod.rs [hunk 1]
```diff
@@ -627,8 +627,14 @@ impl World {
...
+            Some(id) => id,
+            None => return false,
+        };
+        let column = match self.get_populated_resource_column(component_id) {
+            Some(col) => col,
+            None => return false,
+        };
         // SAFE: resources table always have row 0
         let ticks = unsafe { column.get_ticks_unchecked(0) };
         ticks.is_added(self.last_change_tick(), self.read_change_tick())
```
Comment (DJMcNab) [inline crates/bevy_ecs/src/world/mod.rs:640]: This *could* be combined with the line using [`Option::and_then`](https://doc.rust-lang.org/std/option/enum.Option.html#method.and_then), avoiding one match.

I would also change it to be an `if let Some(...)` with an `else {false}` at the end
I.e. my final version would be:
```suggestion
        if let Some(column) = self.components.get_resource_id(TypeId::of::<T>())
             .and_then(|component_id| self.get_populated_resource_column(component_id)) {
            // SAFE: resources table always have row 0
            let ticks = unsafe { column.get_ticks_unchecked(0) };
            ticks.is_added(self.last_change_tick(), self.read_change_tick())
        } else {
            false
        }
```

**Review by DJMcNab — COMMENTED** [review]
I'd implement the function differently, but in a semantically equivalent way. The test looks good though

I'm not sure about `is_resource_changed`. I'd prefer returning `false` when the resource doesn't exist to panicing, but I'm less sure about changing the return type.

File: crates/bevy_ecs/src/world/mod.rs (hunk 1 above)
Comment (RustyStriker) [reply crates/bevy_ecs/src/world/mod.rs:640]: it seems a bit less readable imo, the if condition stretches over 2 lines(or occupies a really long line) and generally seems a bit crowded, the match cases(while seems a bit long) gives you a clear view of what is going on at a glance.
(and generally speaking i would get that if condition to be a variable and a short `if let` statement instead)

**Pull request closed without merging**